### ISSUE TITLE
Stage 2 : Training the Importance Map

### ISSUE THREAD

**Issue by Yuval-H** [adityassrana/Content-Weighted-Image-Compression#2]
First, thanks for the code!
Second, do you plan to add the code to the second stage(training and using the importance map) soon? Do you have a version you can share right now?
Thanks in advance

**Comment by adityassrana** [comment]
Hi, I'm glad you found the code useful. The reason for not uploading it was that when I ported my code from PyTorch to PyTorch Lightning, I didn't get time to port the training script for the importance map. If you want I can add it in the coming month. Until then, I can briefly tell you how to implement it. 

You only have to modify the training step in the script and add an extra entropy error along with the reconstruction error. Have a look at Figure1 and Eqns. 8 - 11 in the paper. and use the Quantizer and Mask already available in `models.py`. 
1. Use the Encoder in `models.py` with `return_imp_map` argument set to True, which will also generate the importance map for you, along with the main feature map.
2. Process this importance map using Quantizer and Mask modules (use the MODULE.apply() method)  (also available in `models.py`)
3. Multiply the processed importance map with the feature map to get the codes to be sent to the decoder.
4. Use Eq. 11 to calculate the entropy error from the importance map (unproccessed by Quantizer and Mask) generated in the previous steps.
5. Sum the reconstruction and entropy loss as in Eq. 8 using gamma = 1e-4?(you'll have to play around with this hyperparameter as I remember the ones mentioned in the paper did not work right away. Use this loss to train the network

I'll try to add the code by next month once :) If you do it before me, I would be happy to receive a pull request.

**Comment by DorianMood** [comment]
It would be really nice if you share the codes you have with me. I coded importance map and tried to train, but the model doesn't converge.

Also, are `Quantizer` and `Mask` models implemented fully? I mean backprop.

**Comment by adityassrana** [comment]
Hi @DorianMood , both `Quantizer`  and `Mask` I've implemented are differentiable. and can be used directly to obtain the importance map. You have to train the main branch of the model before training the importance map. Also, regarding how to train, try following point 5. that I've mentioned above. Even from my experiments, it was very difficult to make the importance map branch converge. I would suggest that you read the papers that are follow-ups to work by the same author for hints as they reuse this idea multiple times. It's a pity they have never released any working code for any of their projects (the caffe files they provide in their repo are incomplete; I know because I tried running their code), so I had to write all of this from scratch while making a lots of guesses  on how to make it work.

If you're going to work on image compression, I would recommend starting from `https://github.com/tensorflow/compression`, as the code is publicly available and the authors are very responsive and helpful. Also, their results are much more promising than this approach.

**Comment by amarzullo24** [comment]
Hi @adityassrana,
Thanks for this work which is a good starting point for my new project!

Related to this issue, I wonder why [the gradients form the mask](https://github.com/adityassrana/Content-Weighted-Image-Compression/blob/master/models.py#L206) are computed using `torch.ones`, which is quite different from equation (7) in the paper. Am I missing something here?

Thanks!

**Comment by aalex1945** [comment]
@DorianMood Hi. Have you finished the code of training importance map yet? I plan to work on the relavant project but I have trouble complement this part. And, have you found out the reason why it dosent converage? Thanks!
